--- ws_2500/Build/bootloader/uboot/common/bootm.c	2019-07-23 08:27:53.000000000 +0530
+++ uboot/common/bootm.c	2020-10-05 17:12:45.910060642 +0530
@@ -835,7 +835,13 @@
 	switch (genimg_get_format(buf)) {
 #if defined(CONFIG_IMAGE_FORMAT_LEGACY)
 	case IMAGE_FORMAT_LEGACY:
-		printf("## Booting kernel from Legacy Image at %08lx ...\n",
+		/* Preventing ## symbols printed during boot. For IPV6 Conformance, 
+		 * the scripts will always be trying for command prompt(#) 
+		 * during reboots by giving enter to proceed further commands with the script.
+		 * During this if BMC is rebooting and if enter is issued from the script when trying for command prompt,
+		 * it will consider these hash prints as command prompt and will try to issue some command which won't work.
+		 * This will result in internal error */
+		printf("** Booting kernel from Legacy Image at %08lx ...\n",
 		       img_addr);
 		hdr = image_get_kernel(img_addr, images->verify);
 		if (!hdr)
